<BR>
<BR>
<BR>
<BR>
<BR>
<BR>
   
                                          
                                          
                                          
                                          
                                          
#                                                 Regression with Abalone Data Set: Linear Regression Model




##                                                       Gladys Murage

##              College of Business, Engineering, and  Technology, National University

##                                         DDS8555 v1: PREDICTIVE ANALYSIS(3602869492)

##                                                        Dr MOHAMED NABEEL

##                                                            March 06, 2025


<BR>
<BR>
<BR>
<BR>
<BR>
<BR>
<BR>
<BR>

# 1. Linear Regression Model
This code performs a linear regression task on a dataset to predict a target variable (Rings in this case) and evaluates the model using the Root Mean Squared Logarithmic Error (RMSLE). It also prepares a submission file for predictions on the test data.


In [102]:
# Load libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_log_error
from sklearn.preprocessing import StandardScaler

# Load the datasets
train_data = pd.read_csv("train.csv")
test_data = pd.read_csv("test.csv")
sample_submission = pd.read_csv("sample_submission.csv")

# Display the first few rows of each dataset
print("Train Data:")
print(train_data.head())

print("\nTest Data:")
print(test_data.head())

print("\nSample Submission:")
print(sample_submission.head())

# Check column names and data types
print("\nTrain Data Columns:", train_data.columns)
print("Train Data Types:\n", train_data.dtypes)

# Preprocess the data
# Handle missing values (if any)
train_data = train_data.dropna()
test_data = test_data.dropna()

# Replace "target_column" with the actual target column name in this care "Rings"
target_column = "Rings"  
# Strip column names to remove extra spaces
train_data.columns = train_data.columns.str.strip()
test_data.columns = test_data.columns.str.strip()

# Encode categorical variables (e.g., 'Sex') in this case
categorical_columns = ["Sex"]  
for col in categorical_columns:
    if col not in train_data.columns:
        raise ValueError(f"Column '{col}' not found in the dataset.")

train_data = pd.get_dummies(train_data, columns=categorical_columns, drop_first=True)
test_data = pd.get_dummies(test_data, columns=categorical_columns, drop_first=True)

# Ensure train and test data have the same columns
train_data, test_data = train_data.align(test_data, join='outer', axis=1, fill_value=0)

# Separate features (X) and target (y) in the training data
X = train_data.drop(columns=[target_column])
y = train_data[target_column]

# Split the training data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Scale the numeric features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)

# Initialize and train the linear regression model
model = LinearRegression()
model.fit(X_train_scaled, y_train)

# Make predictions on the validation set
y_val_pred = model.predict(X_val_scaled)

# Clip negative predictions to zero
y_val_pred = np.maximum(y_val_pred, 0)

# Calculate RMSLE
def rmsle(y_true, y_pred):
    return np.sqrt(mean_squared_log_error(y_true, y_pred))

rmsle_val = rmsle(y_val, y_val_pred)
print("Validation RMSLE:", rmsle_val)

# Preprocess the test data
X_test = test_data.drop(columns=[target_column])
X_test_scaled = scaler.transform(X_test)

# Make predictions on the test data
y_test_pred = model.predict(X_test_scaled)

# Clip negative predictions to zero
y_test_pred = np.maximum(y_test_pred, 0)

# Prepare the submission file
sample_submission[target_column] = y_test_pred
sample_submission.to_csv("submission.csv", index=False)

print("Submission file saved as 'submission.csv'")

Train Data:
   id Sex  Length  Diameter  Height  Whole weight  Whole weight.1  \
0   0   F   0.550     0.430   0.150        0.7715          0.3285   
1   1   F   0.630     0.490   0.145        1.1300          0.4580   
2   2   I   0.160     0.110   0.025        0.0210          0.0055   
3   3   M   0.595     0.475   0.150        0.9145          0.3755   
4   4   I   0.555     0.425   0.130        0.7820          0.3695   

   Whole weight.2  Shell weight  Rings  
0          0.1465        0.2400     11  
1          0.2765        0.3200     11  
2          0.0030        0.0050      6  
3          0.2055        0.2500     10  
4          0.1600        0.1975      9  

Test Data:
      id Sex  Length  Diameter  Height  Whole weight  Whole weight.1  \
0  90615   M   0.645     0.475   0.155        1.2380          0.6185   
1  90616   M   0.580     0.460   0.160        0.9830          0.4785   
2  90617   M   0.560     0.420   0.140        0.8395          0.3525   
3  90618   M   0.570     0.

# 2. Verify the Submission File
Ensure the submission file matches the format of the sample submission file.

In [144]:
# Print the top 5 records
print(submission.head())

      id      Rings
0  90615   8.511788
1  90616  10.006078
2  90617  10.282228
3  90618  10.932569
4  90619   7.759140


In [146]:
# Print the last 5 records
print(submission.tail())

           id      Rings
60406  151021   6.610125
60407  151022   9.199089
60408  151023  10.789541
60409  151024  13.520669
60410  151025   8.570570


In [148]:
# Print the shape of the submission DataFrame
print("Shape of submission file:", submission.shape)

Shape of submission file: (60411, 2)


# 3. Evaluate the Model 
If one has  access to the true target values for the test data (e.g., in a competition), you can evaluate the model's performance using metrics like Mean Squared Error (MSE) or R-squared (R²). However, in this Kaggle competition  the goal is to predict the age of abalone using various physical measurements. The evaluation metric for this competition is Root Mean Squared Logarithmic Error (RMSLE). The target feature is Rings.
## Key things to note:
1. Clip Predictions: ensure predictions are non-negative before applying the logarithm, as log(0) or log(negative) is undefined.
2. Logarithmic Transformation: RMSLE uses log1p (log of 1 + value) to handle zero values well.

# 4.  Interpretation of the RMSLE Value
The Validation RMSLE (Root Mean Squared Logarithmic Error) value of 0.1668 is a measure of how well the regression model is performing on the validation (test) set. The interpretation of this RMSLE  value is as follows:
## What is RMSLE?
RMSLE is a metric used to evaluate regression models, especially when the target variable has a wide range of values or is skewed. It measures the difference between the predicted and actual values on a logarithmic scale.

## Interpretation of RMSLE = 0.1668
A lower RMSLE is better: RMSLE is a non-negative metric, and a value of 0 would indicate perfect predictions. A smaller RMSLE indicates better model performance.

Scale of RMSLE: RMSLE is on a logarithmic scale, so the interpretation depends on the context of the  data. However, in general:

RMSLE < 0.1: indicates very good model performance.

RMSLE between 0.1 and 0 : indicates decent performance but may need improvement.

RMSLE > 0.2: indicates poor performance and significant room for improvement.

The obtained resulting RMSLE (0.1668) is a moderate value, indicating that the regression model is performing decently on the validation set. The predictions are not close to the actual values on a logarithmic scale. Model improvement is called for.

RMSLE is scale-independent: it is less sensitive to large outliers compared to RMSE.

RMSLE penalizes underestimates more than overestimates due to the logarithmic transformation.
## What Does RMSLE = 0.1668 Mean?
Logarithmic Scale: RMSLE measures the relative error between the predicted and actual values on a logarithmic scale. A value of 0.1668 means that, on average, the logarithmic difference between the predicted and actual values is 0.1668.

Relative Error: RMSLE is less sensitive to large outliers compared to RMSE (Root Mean Squared Error), so it focuses more on the relative error rather than the absolute error. This makes it a good metric for datasets where the target variable spans a wide range or is skewed.

## Context of the regression  Model:
Decent Performance: an RMSLE of 0.1668 suggests that the model is performing reasonably well but is not yet optimal. It is making predictions that are somewhat close to the true values, but there is still room for improvement.

Potential Overfitting or Underfitting: If the training RMSLE is much lower than the validation RMSLE, your model might be overfitting. If both are high, the model might be underfitting.
## Why is the RMSLE = 0.1668?
Here are some potential reasons for this RMSLE value:

## Model-Related Issues
### Underfitting:

The model might be too simple in this case; Linear Regression to capture the underlying patterns in the data.

Solution: to apply more complex models like Random Forest, Gradient Boosting, XGBoost, LightGBM, or Neural Networks.

### Overfitting:

The model might be too complex and is memorizing the training data instead of generalizing to unseen data.

Solution: Use regularization techniques (e.g., L1/L2 regularization) or simplify the model.

Poor Feature Engineering:

The features used might not be sufficiently informative or might need transformation (e.g., log transformation, polynomial features).

Solution: Perform feature engineering to create more meaningful features.

### Hyperparameter Tuning:

The model's hyperparameters might not be optimized.

Solution: use techniques like Grid Search or Random Search to find the best hyperparameters.

## Data-Related Issues
### Outliers:

Outliers in the target variable can skew the model's predictions.

Solution: identify and handle outliers (e.g., clipping, transformation).

### Skewed Target Variable:

If the target variable (Rings) is highly skewed, the model might struggle to predict extreme values accurately.

Solution: apply a log transformation to the target variable to reduce skewness.

### Missing Values:

Missing values in the features or target variable can affect model performance.

Solution: handle missing values (e.g., imputation, removal).

### Insufficient Data:

If the dataset is small, the model might not have enough data to learn effectively.

Solution: collect more data or use data augmentation techniques.
## Next Steps
Check for overfitting to ensure that the model is not overfitting the training data. A comparison of the training RMSLE and validation RMSLE will be conducted. If the training RMSLE is much lower than the validation RMSLE, the model may be overfitting.

Test Set Evaluation: since there is a separate test data set, I will evaluate the model to confirm its performance on completely unseen data.

Model Improvement: in order to further improve the model, I would consider the following:

Feature engineering.

Hyperparameter tuning.

Trying more advanced models e.g., Random Forest and  Gradient Boosting. I will conduct both.

# 5. Comparing Training and Validation RMSLE
To check for overfitting, calculate the RMSLE on the training set and compare it to the validation RMSLE
## How to Improve the Model
Step 1: Compare Training and Validation RMSLE
Check if the model is overfitting or underfitting by comparing the training and validation RMSLE:

If Training RMSLE << Validation RMSLE: The model is overfitting.

If Training RMSLE ≈ Validation RMSLE: The model is underfitting.

In [42]:
# Import library
import numpy as np

def rmsle(y_true, y_pred):
    # Ensure there are no non-positive values in y_true and y_pred
    mask = (y_pred > 0) & (y_true > 0)
    y_true_filtered = y_true[mask]
    y_pred_filtered = y_pred[mask]
    
    # Calculate the RMSLE
    return np.sqrt(np.mean(np.power(np.log1p(y_pred_filtered) - np.log1p(y_true_filtered), 2)))

# Calculate and print metrics using the new RMSLE function
train_rmsle = rmsle(y_train, y_train_pred)
val_rmsle = rmsle(y_val, y_val_pred)

print(f"Training RMSLE: {train_rmsle}")
print(f"Validation RMSLE: {val_rmsle}")


Training RMSLE: 0.1624395498468596
Validation RMSLE: 0.16491523791503435


# Comparison of Training and Validation RMSLE results:
Training RMSLE ≈ Validation RMSLE: The training and validation RMSLE values are very close, which suggests that the model is generalizing well and is not overfitting or underfitting significantly.

## Small Gap: 
The small difference between the two values (0.1624 vs. 0.1649) indicates that the model is performing consistently on both the training and validation sets.

## What Does this Mean for the linear regression Model?
Good Generalization: the model is not overfitting (memorizing the training data) or underfitting (failing to capture patterns in the data). It is performing well on both the training and validation sets.

Consistent Performance: the small gap between the training and validation RMSLE suggests that the model is robust and reliable.

Room for Improvement: while the model is performing well, there is still room to improve the RMSLE further. An RMSLE of 0.1649 is decent but not optimal. This model is most likely underfitting.
## How to Improve the Model
Step 1: try a More Complex Model
If the model is underfitting, try a more complex model like Random Forest, Gradient Boosting, or XGBoost. I will attempt a random forest and if that still does not differentiate the RMSLE I will create an XG Boost model.